### PR TITLE
refactor(option.c): misc

### PR DESCRIPTION
**Pull request by lewis6991**
Continued work on deconstructing the house of cards that it option setting.

### CODE REVIEW

File: src/nvim/option.c [hunk 1]
```diff
@@ -2927,181 +2927,188 @@ static const char *check_num_option_bounds(long *pp, long old_value, long old_Ro
...
+  }
+
+  // Save the global value before changing anything. This is needed as for
+  // a global-only option setting the "local value" in fact sets the global
+  // value (since there is only one value).
+  if ((opt_flags & (OPT_LOCAL | OPT_GLOBAL)) == 0) {
+    old_global_value = *(long *)get_varp_scope(&(options[opt_idx]), OPT_GLOBAL);
+  }
+
+  errmsg = validate_num_option(pp, value);
```
Comment (lewis6991) [inline src/nvim/option.c:3110]: diff is a bit broken, but I've removed a chunk of code with this line.

**Review by justinmk — APPROVED** [review]
cc @famiu

File: src/nvim/optionstr.c [hunk 2]
```diff
@@ -415,7 +415,7 @@ void set_string_option_direct_in_win(win_T *wp, const char *name, int opt_idx, c
 ///
 /// @return NULL on success, an untranslated error message on error.
 const char *set_string_option(const int opt_idx, const char *const value, const int opt_flags,
-                              char *const errbuf, const size_t errbuflen)
+                              int *value_checked, char *const errbuf, const size_t errbuflen)
```
Comment (famiu) [inline src/nvim/optionstr.c:418]: Maybe add a comment for the `value_checked` param. Also, could this be `bool *`? I only see it (and also `os_value_checked`) only be assigned to true/false

**Review by famiu — APPROVED** [review]

**Review by justinmk — APPROVED** [review]

**Review by zeertzjq — APPROVED** [review]

**Comment by clason** [comment]
ASAN (build) failure is consistent, but opaque. Maybe just rebase and retrigger?

**Comment by lewis6991** [comment]
I'm pretty confident it'll be caused by this PR. Would need to bisect.

**Comment by zeertzjq** [comment]
```
=================================================================
==214072==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 1 byte(s) in 1 object(s) allocated from:
    #0 0x556eb82c6919 in __interceptor_malloc (**/build/bin/nvim+0xc5e919) (BuildId: 87802768b6d0339f27a2cfa3f07fa629b901150a)
    #1 0x556eb90a4306 in try_malloc **/src/nvim/memory.c:88:15
    #2 0x556eb90a44dc in xmalloc **/src/nvim/memory.c:122:15
    #3 0x556eb934c94a in stropt_copy_value **/src/nvim/option.c:872:18
    #4 0x556eb934c065 in stropt_get_newval **/src/nvim/option.c:1032:14
    #5 0x556eb93441d6 in do_set_option_string **/src/nvim/option.c:1114:18
    #6 0x556eb9341f10 in do_set_option_value **/src/nvim/option.c:1265:5
    #7 0x556eb92eae87 in do_set_option **/src/nvim/option.c:1413:3
    #8 0x556eb92e4e01 in do_set **/src/nvim/option.c:1463:9
    #9 0x556eb92e4280 in ex_set **/src/nvim/option.c:740:9
    #10 0x556eb8beccde in execute_cmd0 **/src/nvim/ex_docmd.c:1634:7
    #11 0x556eb8bc9cd8 in do_one_cmd **/src/nvim/ex_docmd.c:2293:7
    #12 0x556eb8bba56d in do_cmdline **/src/nvim/ex_docmd.c:592:20
    #13 0x556eb8bbf6c0 in do_cmdline_cmd **/src/nvim/ex_docmd.c:291:10
    #14 0x556eb90a7db9 in free_all_mem **/src/nvim/memory.c:726:3
    #15 0x556eb83266be in os_exit **/src/nvim/main.c:678:3
    #16 0x556eb8324c58 in getout **/src/nvim/main.c:802:3
    #17 0x556eb8c21ac0 in ex_quit **/src/nvim/ex_docmd.c:4580:7
    #18 0x556eb8beccde in execute_cmd0 **/src/nvim/ex_docmd.c:1634:7
    #19 0x556eb8bc9cd8 in do_one_cmd **/src/nvim/ex_docmd.c:2293:7
    #20 0x556eb8bba56d in do_cmdline **/src/nvim/ex_docmd.c:592:20
    #21 0x556eb8bbf6c0 in do_cmdline_cmd **/src/nvim/ex_docmd.c:291:10
    #22 0x556eb832a3f2 in exe_commands **/src/nvim/main.c:1907:5
    #23 0x556eb831094c in main **/src/nvim/main.c:579:5
    #24 0x7fd2a5ac284f  (/usr/lib/libc.so.6+0x2384f) (BuildId: 2f005a79cd1a8e385972f5a102f16adba414d75e)

SUMMARY: AddressSanitizer: 1 byte(s) leaked in 1 allocation(s).
```
Apparently it's caused by the last commit, because previous commits don't fail build.

File: src/nvim/option.c [hunk 3]
```diff
@@ -1110,91 +1111,85 @@ static void do_set_option_string(int opt_idx, int opt_flags, char **argp, int ne
...
-                                    opt_flags, value_checked);
+  // Handle side effects, and set the global value for ":set" on local
+  // options. Note: when setting 'syntax' or 'filetype' autocommands may
+  // be triggered that can cause havoc.
+  *errmsg = did_set_string_option(curbuf, curwin, opt_idx, (char **)varp, oldval,
+                                  errbuf, errbuflen,
+                                  opt_flags, value_checked);
 
-    secure = secure_saved;
-  }
```
[the remark this answers is not in the document]
Comment (lewis6991) [reply src/nvim/option.c:1180]: I assumed it was to just isolate scope of 2 variables. I don't know what other reasons there could be.

[the remark this answers is not in the document]
Comment (famiu) [reply src/nvim/option.c:1180]: > I assumed it was to just isolate scope of 2 variables. I don't know what other reasons there could be.

Just appeared really strange to me, which is why I asked. Generally I don't think it's a good idea to remove code without knowing why it was there but in this case it appears harmless

**Review by famiu — APPROVED** [review]

**Pull request merged**